Issue 598: Refactoring List class

veronikaro opened the pull request:
Solves #598.

veronikaro commented:
@gavv I would appreciate your feedback to this PR. Thanks!

gavv commented:
Hi, thanks for PR!

A few comments:

1. Please fix CI. There is a crash in tests reported by several jobs. (You can reproduce it locally by using --sanitizers option).

1. Some ListImpl methods use ListNodeData, and some use ListNode. For consistency, I suggest to use ListNodeData everywhere.

1. I think we could keep container_of_() in template class, after all ListImpl methods will switch to ListNodeData.

1. ListImpl::is_empty() can be removed. We can implement List::is_empty() using ListImpl.size().

1. We can simplify `~List()` to something like (see below why):

    ```cpp
    while (!is_empty()) {
        pop_front();
    }
    ```

1. Also we can move `impl_.head.list = NULL;` from `~List()` to `~ListImpl()`. Since ListImpl is responsible to initialize head, it's logical that it's also responsible to deinitialize it.

1. After these changes we can make check_is_member() a private method of ListImpl.

1. Since ListImpl is a class with methods, not a C-like struct, for consistency I suggest to make head field private and add a public getter `ListNodeData& head()`.

veronikaro commented:
Hi @gavv, 
sorry for the delay. 
I addressed all your comments, thank you very much for the feedback.

>     8. Since ListImpl is a class with methods, not a C-like struct, for consistency I suggest to make head field private and add a public getter `ListNodeData& head()`.

I did this at the cost of implementing `push_back` as a function of `ListImpl` level. The reason for this is that `push_back` requires a pointer to the head. 

I would appreciate your review!

gavv commented on src/internal_modules/roc_core/list_impl.h:
```diff
@@ -0,0 +1,102 @@
+    //!  @p data must be registered in this list.
+    void remove(ListNode::ListNodeData* data);
+
+    //! Get head of list
+    ListNode::ListNodeData head() const;
```
I guess we should return a reference or a pointer here. It's expected that you can take address of any node in list.

gavv commented on src/internal_modules/roc_core/list_impl.h:
```diff
@@ -0,0 +1,102 @@
+private:
+    size_t size_;
+    //! Head of list
+    ListNode::ListNodeData _head;
```
nit: `_head` => `head_`

gavv commented on src/internal_modules/roc_core/list_impl.h:
```diff
@@ -0,0 +1,102 @@
+    size_t size_;
+    //! Head of list
+    ListNode::ListNodeData _head;
+    //! Check if list node data is registered in this list.
+    static void check_is_member(const ListNode::ListNodeData* data, const ListImpl* list);
```
nit: `check_is_member` => `check_is_member_` (because it's private)

gavv commented on src/internal_modules/roc_core/list.h:
```diff
@@ -158,7 +125,9 @@ class List : public NonCopyable<> {
     //!  @p element should not be member of any list.
     void push_back(T& element) {
-        insert_(element, &head_);
+        impl_.push_back(element.list_node_data());
+
+        OwnershipPolicy<T>::acquire(element);
```
```
impl_.insert(element.list_node_data(), &impl_.head());
```

this should work (if you'll return a reference from head()) - and impl.push_back won't be needed.

gavv reviewed (commented):
Thanks for update! Some final comments.

gavv commented:
@veronikaro Hi, my recent commit (2510edd81dbd8003d1ea0a74cecefcc0622c9534) introduced lots of conflict with this patch, so I ported your code to the new version and finished remaining issues: bef7210761ad2ff063da7ce55e7a974a3dc9c6fd

Thanks!
